Functinality to save/restore snapshots for monoprice platform

etsinko opened the pull request:
## Description:

Added functionality to save/restore snapshots in the `monoprice` platform. This was inspired by a similar functionality in `sonos` platform. 

This PR adds two services to `monoprice`:

- `monoprice_snapshot` - saves zone's volume, source, mute and power status
- `monoprice_restore` - restores zone's values taken previously with `monoprice_snapshot`.

One example use case for this addition is if an user wants to use a monoprice zone for a public announcement. In such a case a snapshot for the zone will be taken first, then the anouncement will be played and finally the zone will be restored to its previous state.

Examples of a service call:
```python
hass.services.call('media_player', 'monoprice_snapshot', 
                                            {'entity_id': 'media_player.zone_1'},
```


**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3874

## Example entry for `configuration.yaml` N/A

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

etsinko committed:
added functionality to save/restore snapshots to monoprice platform

pvizeli commented:
Can you implement the snapshot to `media_player`? I think it is time to handle that central

etsinko commented:
How do you see this? Media player class registers services `snapshot` and `restore` and implements stub methods that do nothing by default?

etsinko committed:
Merge branch 'dev' into monoprice-snapshot

etsinko commented:
@pvizeli I see that there are 3 media players that currently use `snapshot` and `restore` functionality:
- `monoprice`
- `snapcast`
- `sonos`

Both `monoprice` and `snapcast` are pretty much identical and straighforward - just a simple service call, no parameters. `sonos` on the other hand supports an additional parameter passed to the service call - `with_group`.

johnnyletrois commented:
What's the hold up here?  Is there anything I can do to help?  I was hoping this would be merged before Christmas.

etsinko commented:
I also think this needs to be merged first and then we can extract common functionality for `snapshot` and `restore` into the base class.

etsinko committed:
renamed monoprice_snapshot, monoprice_restore to snapshot, restore

This is to simplify refactoring of snapshot/restore functionality for monoprice, snapcast and sonos in the future
